### DISCUSSION TITLE
Does firejail improve the security of my system? thoughts by @rusty-snake

### DISCUSSION THREAD

**Discussion by rusty-snake** [netblue30/firejail#4601] [General]
Previous/related discussions:

* #3046
* #3552
* #4522
* https://github.com/netblue30/firejail/issues/4210#issuecomment-841882340
* #5779
* #6466
* https://github.com/netblue30/firejail/issues/6675#issuecomment-2764702732

Related PRs:

* #5290

-----

*THIS IS MY OPINION*

Virtually all the arguments against firejail can be summed up in "Firejail
worsens security by acting as a privilege escalation hole because it is complex
and requires to be SUID." We all know that complex means bugs and bugs in SUID
programs cause more damage. Normally we would be done here and could say "this
program worsens security". However, since firejail is a program that is
supposed to increase security, maybe the overall balance is positive (i.e. the
security gains outweigh the new risks introduced). But is the overall balance
actually positive or not? Well, it depends on your [thread models].

Privacy Guides has a very good article about Thread Modeling (https://www.privacyguides.org/threat-modeling/)
that defines five questions:

> 1. What do I want to protect?
> 2. Who do I want to protect it from?
> 3. How likely is it that I will need to protect it?
> 4. How bad are the consequences if I fail?
> 5. How much trouble am I willing to go through to try to prevent potential consequences?

I will answer these questions for myself (a graphical Linux system used by only
one user) however, I think that these answers are at least roughly correct for
most of you.

[thread models]: https://en.wikipedia.org/wiki/Threat_model

### 1. What do I want to protect?

The Confidentiality and Integrity of my system. That includes that my password
and keys are kept secret, my emails private, my github account secure and my
pictures/documents/code/... are not being manipulated.

#### 1.1. What do I not want to protect?

The Availability of my system because I break it myself more often than I am
even attacked.

The kernel and the root user. Although their integrity is a prerequisite for
the above points, it is not the goal. Meaning I scare remote>kernel>password
attacks but not remote>firefox>kernel attacks if remote>firefox>password is
possible too.

### 2. Who do I want to protect it from?

Cyber Crooks who loitering on the internet and preying on random victims.

#### 2.1. Who do I *not* want to protect it from?

Intelligence and targeted attacks. If they are in your thread model
use [Tails], [Whonix] or [Qubes OS].

[Tails]: https://tails.boum.org/
[Whonix]: https://www.whonix.org/
[Qubes OS]: https://www.qubes-os.org/

### 3. How likely is it that I will need to protect it?

Cyber Crooks usually attack Windows systems, IoT-the-s-stands-for-security-devices
and Servers.

Such attacks don't use unpublished zero-days. Instead they use publicly known
(and often patched) vulnerabilities or Social Engineering. Therefore install
updates! and don't follow "to install this cool program just `curl | bash`"
instructions.

### 4. How bad are the consequences if I fail?

It depends on the attack but one of the worsts would be that someone does criminal things and I am blamed for it.

### 5. How much trouble am I willing to go through to try to prevent potential consequences?

This is an important point, because you can have too much security. I could use
- Tails, however it has a lot restriction and make thing complicated. Too complicated most of the time.
- Qubes OS, however it has high hardware requirements.
- Bubblewrap, however I would need to write profiles (and even the concept of profiles)
  myself, compile seccomp filters myself, setup D-Bus filtering myself, do the desktop
  integration myself (TBH I did this for firejail) and know a lot of my system and
  the program I want to run. Except of a few experts no one of use can do this correct
  (and no you aren't an expert). You will either have a weak sandbox because you forgot
  things, don't know about things or made mistakes or you will have a strict sandbox, a
  too strict sandbox and give up sandboxing.
- an custom sandboxing program, however this has the same drawbacks a bubblewrap
- AppArmor/SELinux, however this is more complicated but if you have the knowledge, go for it.

## Summary

I am worried that an attacker can execute commands with full user rights.
An attack who execute commands with full user rights is effectively already root. 

![If someone steals my my laptop while I'm logged in, they can read my email, take my money, and impersonate me to my friends, but at least they can't install drivers without my permission.](https://imgs.xkcd.com/comics/authorization.png)
https://xkcd.com/1200/

**Comment by TommyTran732** [comment]
This wasn't my main point in the PG discussion. I was comparing it to Flatpak. The problem with using SUID was just one thing I brought up.

The main reason I argued for Flatpak usage over Firejail is its usability (all Flatpak packages are confined by a set of permissions by default and the user can easily use a GUI like Flatseal to toggle those permissions), integration with the rest of the ecosystem (Portal, GNOME, etc). My ideal setup (especially for a new user who had just come to Linux) is Wayland + Pipewire API for audio + Flatpak (which, again, integrates nicely with Portal) + Flatseal. That way they have some what of a permission control system on their Linux computers. I just don't see how Firejail fits in. I also think modifying firejail profiles and making them would be easier for the user than just changing toggles in Flatseal. Firetools would have been nice if it could actually save the profile it creates for later use.

**Reply by crocket** [reply]
Firejail is a lot more flexible than flatpak if you want to do anything useful at low level.

How do you use piprewire for audio? Most applications don't support pipewire. I use ALSA currently.
I haven't tried piprewire's ALSA plugin because it might add too much latency to audio or crackle if CPU is overloaded.

**Reply by SergSel2006** [reply]
> How do you use piprewire for audio? Most applications don't support pipewire. I use ALSA currently. I haven't tried piprewire's ALSA plugin because it might add too much latency to audio or crackle if CPU is overloaded.

Firstly, you can use pipewire for audio with apications that use Pulseaudio or JACK or ALSA. 
Secondly, the cracling is torelable but it depends on what you want to do and what is your current setup (basic 2 output devices and a mic have zero crackling even under stress loads). I can guarantee at least that it works with realtime privileges, so latency is stable.

**Reply by crocket** [reply]
I moved everything to pipewire. I have used pipewire for months. It doesn't produce crackles if CPU cores are overloaded. Its latency is okay. I like pipewire. Pipewire is okay.

**Comment by topimiettinen** [comment]
For xkcd-1200, Firejail could support using several UIDs (real, not subuids) under control of a main UID. This would of course need to be enabled at system configuration, not by an user. Then simple DAC controls (chmod 0700, groups) could be used to control which UID can access which files. Some folders could be shared (tradeoff of convenience vs. security). For example, Linux firewalls can use different rules based on the UID (NFT: `meta skuid != 1000 drop`).

Previous discussions in #3892, #208.

**Reply by matu3ba** [reply]
@topimiettinen What is the advantage to editing with visudo the sudoers file with `NOPASSWD: /etc/myscripts/do-whatever` to start the configured program? You need to setup the machanism for file sharing or copying between users anyway.

Also: If you want to automatize this brittle stuff, you end up adding the complexity of a full-blown user agent. And the advantage would only help for programs that can run as local user and not escalate to root, which is not too tricky with desktop complexity.
Or you would have them contained in the first place.

Regarding the information leakage: Most of (and all widespread) Linux Desktops are not designed with that in mind, which makes it unfeasible to aim for.

There [may be hope](https://www.gobolinux.org/), but I dont see a strategy to expand this yet. And [hacks seem to be the essence of Linux](https://www.usenix.org/conference/osdi21/presentation/fri-keynote).

**Reply by crocket** [reply]
If you like gobolinux, look into GNU Guix system, too.

**Reply by topimiettinen** [reply]
`ssh localhost` using `authorized_keys` is another option.

The advantage would be that Firejail could take over the UID and related home directory completely. The master version of data would be in the normal user's directory and the sub-UID directories would be just for setting up the stage.

GoboLinux seems to just rename the paths, like NixOS. I don't see much benefit from this, except maybe "security by obscurity". It could still mean that simple metasploit stuff, which expects to find commands in usual places, may fail.

Linux isn't perfect but the alternatives aren't either.

**Comment by topimiettinen** [comment]
I think there were also interesting discussions around [BeEF](https://beefproject.com/) (Browser Exploitation Framework), but I can't find it.

**Reply by SkewedZeppelin** [reply]
Weird. I know exactly the issue you mention, but I cannot find it at all! Tried a few ways too.

**Reply by rusty-snake** [reply]
GitHub hides it in search and in the list, maybe because the author is @ghost. Anyway here is it https://github.com/netblue30/firejail/discussions/4436.

**Comment by crocket** [comment]
For protection against executing arbitrary code with full user rights, firejail is great.
As I once wrote, the combined attack surface of programs that firejail restricts is more than million times larger than that of firejail.

I also think that it is far better than nothing against targeted attacks by intelligence agencies.

But, firejail's security can be greatly improved by splitting out privileged parts as separate binaries.
I think firejail will end up producing multiple binaries if it wants to split out privileged parts.

AppArmor can be combined with firejail, and I can introduce both firejail and AppArmor gradually per each application.

Firejail can protect users from targeted attacks to some degrees because most targeted attacks don't rely on exploits in SUID binaries or kernel. For example, `firejail --net=none mpv` can prevent mpv from fetching media metadata from a server run by an intelligence agency. Intelligence agencies insert their own servers into media metadata so that media players consult their servers for metadata. People usually use media players without tor, so they end up revealing their real IP address and thus their physical location to intelligence agencies. And then, they can send police officers to arrest political dissenters.

**Reply by curiosityseeker** [reply]
> AppArmor can be combined with firejail, and I can introduce both firejail and AppArmor gradually per each application.

Really? Firejail works flawlessly with the `firejail-default` profile but, IMO, not necessarily with specific or stock AppArmor profiles. I had discussed this matter at length [here](https://github.com/netblue30/firejail/issues/3226) with @smitsohu, and my conclusion is that combining AppArmor with Firejail is problematic. I must admit that I haven't tried it since that discussion and I hadn't thoroughly tested @smitsohu 's [proposal](https://github.com/netblue30/firejail/issues/3226#issuecomment-594142656). If you have a working solution I would be highly interested.

**Reply by curiosityseeker** [reply]
> Firejail can protect users from targeted attacks to some degrees because most targeted attacks don't rely on exploits in SUID binaries or kernel. 

And even the danger of exploits of SUID binaries or attacking the kernel is greatly mitigated by Firejail. Executing SUID binaries (including firejail) from within sandboxed application is not possible, and the kernel attack surface is much smaller by the use of seccomp and by dropping capabilities. Since there are now Firejail profiles for many applications, practically all relevant applications are sandboxed. Hence, an attacker must find ways to bypass those sandboxed applications somehow in order to infect a system.

**Reply by crocket** [reply]
Do you think AppArmor is not very useful in the presence of firejail?

**Reply by rusty-snake** [reply]
The general advise is to use either firejail (with `apprmor` aka. `firejail-default`) or AppArmor (e.g. `usr.bin.firefox`) for a program.

**Reply by crocket** [reply]
AppArmor is desirable for login managers, terminal emulators, and shell because they can't be used with firejail which doesn't permit nested sandboxes.

**Reply by smitsohu** [reply]
I think I'll try to upstream [this snippet](https://github.com/netblue30/firejail/issues/3226#issuecomment-594142656) which should resolve incompatibilities with Firejail. Unfortunately it won't have effect in AppArmor  < 3.0, because the base abstraction there doesn't `include` any other path.

**Reply by curiosityseeker** [reply]
> Do you think AppArmor is not very useful in the presence of firejail?

Yes, it is. But it shouldn't be used with Firejail for the same applications (as @rusty-snake already said).

**Reply by curiosityseeker** [reply]
> > Do you think AppArmor is not very useful in the presence of firejail?
> 
> Yes, it is. But it shouldn't be used with Firejail for the same applications (as @rusty-snake already said).

Okay, that statement was a bit too hasty. I've finally tested again @smitsohu  's snippet mentioned above with Firefox - and it seems to work (after adding `ignore apparmor` to `firefox.local`, of course). As mentioned in that snippet, profile transitions do not work: I have an AppArmor profile for okular, and so far PDFs in Firefox were opened in okular via `/usr/bin/okular rPx,`. This is no longer possible - so I switched back to the built-in pdf.js which [has become much better and more secure](https://hacks.mozilla.org/2021/10/implementing-form-filling-and-accessibility-in-the-firefox-pdf-viewer/) (I guess that Fission should help here, too) . That's a price I'm willing to pay. This will be probably more difficult to implement for other applications which are in need of some helper applications, though.

**Reply by smitsohu** [reply]
@curiosity-seeker Maybe there's a way to make profile transitions work. If I understand it right it involves an AppArmor profile for Firejail itself (the SUID Firejail) that allows _everything_. No restrictions whatsover. Then the kernel should remember that nonewprivs was set with this profile enforced and compare all profile transitions against it. If that doesn't work we probably need to modify Firejail so that it accepts an AppArmor profile as argument and stacks that profile on the currently enforced one.

**Reply by smitsohu** [reply]
If I understand it right there is work going on to provide a new set of transitions in AppArmor to deliberately ignore nonewprivs. So at some point there will be a proper solution, too.

**Reply by crocket** [reply]
I think it's simpler to just use firejail alone for an application and apparmor for applications that are going to launch other applications.

Login manager and shell need to launch other applications.

**Reply by topimiettinen** [reply]
FYI: AppArmor's problems don't exist with SELinux: if a domain transition is allowed, it doesn't matter after that what was the original domain. So firejailed games can transition from `user_u:user_r:user_t:s0` to `user_u:user_r:games_t:s0`, Firefox to `user_u:user_r:mozilla_t:s0` etc. The same rules apply to the new domain, with or without Firejail.

**Reply by crocket** [reply]
I don't understand what you mean because I haven't actually used apparmor or selinux. Can you say in plain english?

**Reply by curiosityseeker** [reply]
> FYI: AppArmor's problems don't exist with SELinux: if a domain transition is allowed, it doesn't matter after that what was the original domain. 

Interesting! On the other hand, SELinux hardly confined any applications but rather system processes the last time I had a look at it. I don't know if that has changed in the meantime.

**Reply by rusty-snake** [reply]
That depends on the policy you use. Nothing stops you from doing so, but you are right that the policies used by e.g. Fedora run mostly all user programs unconfined.

**Reply by curiosityseeker** [reply]
> That depends on the policy you use. Nothing stops you from doing so, but you are right that the policies used by e.g. Fedora run mostly all user programs unconfined.

Yeah, [this](https://danwalsh.livejournal.com/72697.html) was the post by SELinux pope Dan Walsh I had in mind. However, in Gentoo (Hardened?) it's done [differently](https://blog.siphos.be/2015/08/why-we-do-confine-firefox/). And there is the [SELinux](https://danwalsh.livejournal.com/28545.html) [sandbox](https://blog.agchapman.com/selinux-browser-sandbox/), of course.

**Reply by curiosityseeker** [reply]
@smitsohu 
> Maybe there's a way to make profile transitions work. If I understand it right it involves an AppArmor profile for Firejail itself (the SUID Firejail) that allows _everything_. No restrictions whatsover. 

Hm, it seems that this does not work. I created a profile which looks like this:
```
# Last Modified: Wed Oct 20 14:47:02 2021
abi <abi/3.0>,

include <tunables/global>

/usr/bin/firejail {
  capability chown,
  capability dac_override,
  capability dac_read_search,
  capability fowner,
  capability mknod,
  capability setgid,
  capability setpcap,
  capability setuid,
  capability sys_admin,
  capability sys_chroot,
  capability sys_ptrace,

  network inet dgram,
  network inet stream,
  network inet6 dgram,
  network inet6 stream,
  network netlink raw,
  network unix dgram,
  network unix seqpacket,
  network unix stream,

  signal receive set=exists peer=/usr/bin/firejail,
  signal receive set=kill peer=/usr/bin/firejail,
  signal receive set=term peer=/usr/bin/firejail,
  signal send set=exists peer=/usr/bin/firejail,
  signal send set=kill peer=/usr/bin/firejail,
  signal send set=term peer=/usr/bin/firejail,

  ptrace read peer=/usr/bin/firejail,
  ptrace readby peer=/usr/bin/firejail,
  ptrace readby peer=prelockd,
  ptrace trace peer=/usr/bin/firejail,
  
  / r,
  /** mrwlk,
  /** ux,

}
```
Further tests are necessary but right now it doesn't look as if something has changed.

**Reply by topimiettinen** [reply]
> I don't understand what you mean because I haven't actually used apparmor or selinux. Can you say in plain english?

Firejail can confine Firefox by using a mount namespaces so that for example, it can only access ~/Downloads but no other files. But setting up the mounts so that Firefox can open pdfs by starting Okular would loosen the sandbox, because the mount sandbox in Firejail needs to be sum of both apps' needs. With AppArmor (and SELinux), it's possible to confine both programs separately, for example Okular doesn't need access to network but Firefox obviously does. Based on the comments above, AppArmor has problems when Firejail is also involved, but it's doable with SELinux. For me it's OK if Firefox can't execute anything, I can open the downloaded files without it.

**Reply by topimiettinen** [reply]
> That depends on the policy you use. Nothing stops you from doing so, but you are right that the policies used by e.g. Fedora run mostly all user programs unconfined.

The first step to harden SELinux policy is to change from unconfined user `unconfined_u` to unprivileged user `user_u` (optionally also switch root to `sysadm_u`) and `restorecon` file systems. The next step is to fix "a few" bugs in policy. Still, TOMOYO doesn't have any policies readily available and AppArmor profiles are also pretty lacking, so nothing new here.

**Comment by dm17** [comment]
Wouldn't running browsers in docker or podman be even better?

**Reply by topimiettinen** [reply]
Docker etc. use the same technologies as Firejail: mount/PID/user/net namespaces, seccomp, capabilities etc. If they have cool ideas, maybe those could be used also in Firejail.

**Reply by dm17** [reply]
Relevant: https://crlf.link/log/entries/211008-1/

**Reply by rusty-snake** [reply]
If it is done right!, it can have the following benefits:
- a isolated network namespace (admittedly firejail has `net eth1`)
- a private base system
- runsc (aka gVisor)

**Reply by topimiettinen** [reply]
VM technology (gVisor or direct KVM) would be nice in Firejail. It would allow deep inspection of system call arguments without TOCTOUs, because it's possible to "stop the world" while doing so. Mangling the arguments would be also possible, which isn't allowed with seccomp.

**Comment by curiosityseeker** [comment]
> Virtually all the arguments against firejail can be summed up in "Firejail worsens security by acting as a privilege escalation hole because it is complex and requires to be SUID." We all know that complex means bugs and bugs in SUID programs cause more damage. Normally we would be done here and could say "this program worsens security".

Shouldn't it be mentioned that the risk of Firejail being a SUID can be [mitigated](https://firejail.wordpress.com/documentation-2/basic-usage/#suid)?

**Comment by crocket** [comment]
By the way, if intelligence agencies are your threat model, you should flash coreboot on a computer without CPU backdoors.
CPU backdoors can bypass all security measures.
https://freundschafter.com/research/system-alternatives-without-intel-me-iamt-and-amd-psp-secure-technology/

BadBIOS utilizes speakers and microphones as network interfaces because sound is a medium for wireless communication.

All of that cannot be blocked by virtual machines or firejail. Those highly privileged backdoors have direct access to RAM.

As long as CPU backdoors and other highly-privileged backdoors like BadBIOS remain as the lowest hanging fruits, finding exploits in firejail or virtual machines doesn't make much sense.

Intelligence agencies already got you by the balls with CPU backdoors and other highly privileged backdoors in UEFI and BIOS.

**Reply by dm17** [reply]
Right... And it is a pain that no modern/fast CPUs can verifiably disable the "management engines" except for IBM Power9, of course... But it seems Power10 won't even be FOSS (so I see no reason to invest in their significantly pricier hardware).

**Reply by crocket** [reply]
Do you know any combination of CPU and mainboard that supports coreboot and 16~32GB ECC RAM and has no CPU backdoor?

Although my AMD CPU doesn't have AMD PSP, my AM3+ socket mainboard doesn't seem to support coreboot.

**Reply by dm17** [reply]
https://www.raptorcs.com/content/base/products.html

Yes, but how do you verify the lack of PSP? Have to trust their closed software saying the PSP is off / disabled it seems.

Well, let's say "one that we could verify the firmware / lack of black box OS running in CPU"... Only the power9 + ECC as far as I know. It seems more effort to harden these rarer CPUs due to having fewer eyeballs on them. It would be great if we could pressure AMD to release their source - because their modern CPUs do have great security features (but they're black boxes with themselves as root-of-trust).

**Reply by crocket** [reply]
AMD FX-8300 doesn't come with AMD PSP according to
* https://freundschafter.com/research/system-alternatives-without-intel-me-iamt-and-amd-psp-secure-technology/ 
* https://freundschafter.com/research/amd-processors-without-amd-psp-secure-technology/

**Reply by dm17** [reply]
The problem with that is: what's the difference between AMD saying it doesn't come with a PSP and the AMD systems with BIOSs that can set PSP to disabled? Not much because you can't verify it actually being off on either. That's the point of these libre/coreboot processes.

**Reply by crocket** [reply]
If an AMD CPU has AMD PSP, coreboot isn't going to do much because nobody has found a way to (partially) disable AMD PSP. But, coreboot uses me_cleaner to disable Intel ME for older machines.

The global ruling structure has a habit of telling people what they are doing. I think they are honest about lack of AMD PSP on https://freundschafter.com/research/amd-processors-without-amd-psp-secure-technology/

They are honest in such cryptic ways as telling people lack of CPU backdoors in technical documents. Most people are not going to read technical documents on CPU architectures.

AMD PSP has to be documented in some ways for developers.

**Reply by dm17** [reply]
Right. But I don't know much about me_cleaner... The only way to verify lack of backdoors in these closed source firmwares is to bootstrap the CPU with open firmware... Which we can't do on any modern AMD or Intel CPUs - can you? Fastest AMD CPU without PSP is supposedly the Opteron? Still pretty old... :(

Another aspect of running with open firmware is being able to patch your own systems in the case of a cyber attack. A community could collaborate. In the case that you have to rely on AMD/Intel - the speed to react is way larger, the inability to have more people testing and looking through the code sucks, and the ability to turn off some 'network-enabled management engine' purposeful/accidental backdoor is vastly diminished.

I get what you're saying about the way they reveal things, but open firmware is the only verifiable way, isn't it?

**Reply by crocket** [reply]
Presence of AMD PSP seems detectable with some programs. There are programs that can analyze AMD PSP, but nobody has found a way to disable it.

**Reply by dm17** [reply]
Yep... And it seems Power10 will unfortunately have binary blobs until RaptorCS can get the help / resources to remove them somehow. They say it will be expensive. Power9 seems it will be decent for a while longer.

**Reply by crocket** [reply]
Power9 is expensive.

**Reply by amano-kenji** [reply]
The closest you can get to a computer without hardware backdoor is intel CPU computers from star labs, system76, and purism.

Their intel CPU computers come with coreboot that disables Intel ME during boot. They also stripped Intel ME down to the minimum size possible.
